### PR TITLE
Sign using keys stored in hashicorp vault

### PR DESCRIPTION
**Pull request by rain-on**

### CODE REVIEW

File: core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/ArtifactSignerFactoryTest.java [hunk 1]
```diff
@@ -39,14 +42,23 @@
   private Path passwordFile;
   private ArtifactSignerFactory artifactSignerFactory;
 
+  final Vertx vertx = Vertx.vertx();
```
Comment (usmansaleem) [inline core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/ArtifactSignerFactoryTest.java]: shouldln't this be in `@BeforeEach`? as you are closing it in `@AfterEach`?

File: acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java [hunk 2]
```diff
@@ -128,6 +133,34 @@ public void unusedFieldsInRequestDoesNotAffectSigning() throws Exception {
     assertThat(response.getStatusCode()).isEqualTo(HttpResponseStatus.OK.code());
   }
 
+  @ParameterizedTest
+  @ValueSource(strings = {"/signer/block"}) // , "/signer/attestation"})
```
Comment (jframe) [inline acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java]: We want to test this for all endpoints i.e. block, attestation, randao_reveal not just block

File: acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java [hunk 3]
```diff
@@ -128,6 +133,34 @@ public void unusedFieldsInRequestDoesNotAffectSigning() throws Exception {
...
+  public void ableToSignUsingHashicorp(final String artifactSigningEndpoint)
+      throws ExecutionException, InterruptedException {
+    final String configFilename = keyPair.publicKey().toString().substring(2);
+    final DockerClientFactory dockerClientFactory = new DockerClientFactory();
+    final HashicorpNode hashicorpNode =
+        HashicorpNode.createAndStartHashicorp(dockerClientFactory.create(), true);
```
Comment (jframe) [inline acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java]: Do we need to shutdown hashicorp node or that happening automatically somewhere?

File: core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    signingMetadata.put("keyPath", "/v1/secret/data/secretPath");
+    signingMetadata.put("token", "accessToken");
+    signingMetadata.put("tlsEnabled", "true");
+    signingMetadata.put("tlsKnownServersPath", malformedknownServers.toString());
+    YAML_OBJECT_MAPPER.writeValue(filename.toFile(), signingMetadata);
+
+    final Optional<ArtifactSigner> signer = signerProvider.getSigner(PUBLIC_KEY);
+    assertThat(signer).isEmpty();
+
+    assertThat(logAppender.getLogMessagesReceived().size()).isNotZero();
```
Comment (jframe) [inline core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java]: nit: assertThat(logAppender.getLogMessagesReceived()).isNotEmpty() instead

File: core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      signingMetadata.put("keyPath", "/v1/secret/data/secretPath");
+      signingMetadata.put("token", "accessToken");
+      signingMetadata.put("tlsEnabled", "true");
+      signingMetadata.put("tlsKnownServersPath", missingFingerprintFile.toString());
+      YAML_OBJECT_MAPPER.writeValue(filename.toFile(), signingMetadata);
+
+      configsDirectory.toFile().setWritable(false);
+      final Optional<ArtifactSigner> signer = signerProvider.getSigner(PUBLIC_KEY);
+      assertThat(signer).isEmpty();
+
```
Comment (jframe) [inline core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java]: Thinking this test should check the error logs given it's name

File: core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    final Map<String, String> signingMetadata = new HashMap<>();
+    signingMetadata.put("type", "hashicorp");
+    signingMetadata.put("serverHost", "localhost");
+    signingMetadata.put("keyPath", "/v1/secret/data/secretPath");
+    signingMetadata.put("token", "accessToken");
+    signingMetadata.put("tlsEnabled", "false");
+    YAML_OBJECT_MAPPER.writeValue(filename.toFile(), signingMetadata);
+
+    final Optional<ArtifactSigner> signer = signerProvider.getSigner(PUBLIC_KEY);
+    assertThat(signer).isEmpty();
```
Comment (jframe) [inline core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java]: Similar to above, this test implies we are checking the error message but there is not assertion for the error message

File: core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/YamlSignerParserTest.java [hunk 7]
```diff
@@ -105,8 +104,7 @@ void unencryptedMetaDataInfoWithInvalidHexEncodingForPrivateKeyFails() throws IO
 
     assertThatThrownBy(() -> signerParser.parse(filename))
         .isInstanceOf(SigningMetadataException.class)
-        .hasMessageStartingWith(
-            "Invalid signing metadata file format: Invalid hex value for private key");
+        .hasMessageStartingWith("Invalid signing metadata file format");
```
Comment (jframe) [inline core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/YamlSignerParserTest.java]: Is this detail included somewhere else now? Do we need to remove it? I remember we went to the trouble of getting this appearing nice in the exception message.

File: core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/ArtifactSignerFactoryTest.java (hunk 1 above)
Comment (rain-on) [reply core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/ArtifactSignerFactoryTest.java]: done

File: core/src/main/java/tech/pegasys/eth2signer/core/multikey/DirectoryBackedArtifactSignerProvider.java [hunk 8]
```diff
@@ -160,4 +160,17 @@ public ArtifactSigner getSigner() {
       return signer;
     }
   }
+
+  private void renderException(final Throwable t, final String filename) {
+    LOG.error("Error parsing signing metadata file {}: {}", filename, getRootCauseMessage(t));
```
[the remark this answers is not in the document]
Comment (rain-on) [reply core/src/main/java/tech/pegasys/eth2signer/core/multikey/DirectoryBackedArtifactSignerProvider.java]: done

File: acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java (hunk 3 above)
Comment (rain-on) [reply acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java]: done

File: core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java (hunk 4 above)
Comment (rain-on) [reply core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java]: done

File: core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/YamlSignerParserTest.java (hunk 7 above)
Comment (rain-on) [reply core/src/test/java/tech/pegasys/eth2signer/core/multikey/metadata/YamlSignerParserTest.java]: I'll create a test at the Integration Level to show these errors still appear nicely - but they are constructed in the DirectoryBackedSignerProvider - not the YamlSigner

File: acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java (hunk 2 above)
Comment (rain-on) [reply acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java]: done

File: core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java (hunk 6 above)
Comment (rain-on) [reply core/src/integrationTest/java/tech/pegasys/eth2signer/core/DirectoryBackedArtifactSigningProviderIntegrationTest.java]: done

File: acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java (hunk 3 above)
Comment (jframe) [reply acceptance-tests/src/test/java/tech/pegasys/eth2signer/tests/KeyLoadAndSignAcceptanceTest.java]: can't see where the shutdown is being called

**Pull request merged**